### PR TITLE
[RFC] Another attempt at Provision TEE threads for system invocation

### PR DESCRIPTION
**Pull request by jenswi-linaro**
An alternative to what is proposed in https://github.com/linaro-swg/linux/pull/109

@etienne-lms feel free to use these patches if they make sense to you.

### CODE REVIEW

File: drivers/tee/optee/call.c [hunk 1]
```diff
@@ -53,15 +70,40 @@ void optee_cq_wait_init(struct optee_call_queue *cq,
...
+	} else {
+		list_add_tail(&w->list_node, &cq->normal_waiters);
+		if (cq->total_thread_count) {
+			/*
+			 * Claim a normal thread if one is available, else
+			 * we'll need to wait for a normal thread to be
+			 * released.
+			 */
+			if (cq->free_normal_thread_count > 0)
+				cq->free_normal_thread_count--;
```
Comment (etienne-lms) [inline drivers/tee/optee/call.c:90]: some thread protection around this incrementation?

File: drivers/tee/optee/call.c [hunk 2]
```diff
@@ -104,6 +158,9 @@ void optee_cq_wait_final(struct optee_call_queue *cq,
 	/* Get out of the list */
 	list_del(&w->list_node);
 
+	if (!w->sys_thread)
+		cq->free_normal_thread_count++; /* Release a normal thread */
```
Comment (etienne-lms) [inline drivers/tee/optee/call.c:162]: mutex?

**Review by etienne-lms — COMMENTED** [review]
Looks good to me. It is a nice way to let system sessions simply always reach TEE.
Some questions on missing mutex locking.

There is no fallback if an earlier execution stage left a suspended tee thread open. Worst case effect would be regular sessions will not reach OP-TEE, in case sys_thread_cnt == total_thread_cnt - lost_thread_cnt. Should it be addressed? A warning in `optee_smc_do_call_with_arg()`?
```patch
 		if (res.a0 == OPTEE_SMC_RETURN_ETHREAD_LIMIT) {
+			WARN_ONCE(!w->sys_thread, "TEE thread count mismatch\n")

 			/*
 			 * Out of threads in secure world, wait for a thread
 			 * become available.
 			 */
 			optee_cq_wait_for_completion(&optee->call_queue, &w);
 		} else if (OPTEE_SMC_RETURN_IS_RPC(res.a0)) {
```

File: drivers/tee/optee/call.c (hunk 1 above)
Comment (jenswikl) [reply drivers/tee/optee/call.c:90]: We're holding `cq->mutex`.

File: drivers/tee/optee/call.c (hunk 2 above)
Comment (jenswikl) [reply drivers/tee/optee/call.c:162]: We're holding `cq->mutex`.

**Comment by jenswi-linaro** [comment]
> There is no fallback if an earlier execution stage left a suspended tee thread open. Worst case effect would be regular sessions will not reach OP-TEE, in case sys_thread_cnt == total_thread_cnt - lost_thread_cnt. Should it be addressed? 

The worst case is a total deadlock in the secure world if threads are left suspended, that's basically a DOS by starvation of CPU.

> A warning in `optee_smc_do_call_with_arg()`?
> 
> ```diff
>  		if (res.a0 == OPTEE_SMC_RETURN_ETHREAD_LIMIT) {
> +			WARN_ONCE(!w->sys_thread, "TEE thread count mismatch\n")
> 
>  			/*
>  			 * Out of threads in secure world, wait for a thread
>  			 * become available.
>  			 */
>  			optee_cq_wait_for_completion(&optee->call_queue, &w);
>  		} else if (OPTEE_SMC_RETURN_IS_RPC(res.a0)) {
> ```

Yes, it makes sense.

File: drivers/tee/optee/call.c (hunk 1 above)
Comment (etienne-lms) [reply drivers/tee/optee/call.c:90]: right, sorry.

**Comment by jenswi-linaro** [comment]
Adding the `WARN_ONCE`

**Comment by etienne-lms** [comment]
I've posted **the** changes to LKML: [system session patches v5](https://lore.kernel.org/lkml/20230320162801.2811542/).
Note I removed the call to `WARN_ONCE()` I suggested. I found the backtrace information printed does not add much value.

**Comment by etienne-lms** [comment]
Patch v6 posted, see https://lore.kernel.org/lkml/20230505173012.881083 series.

**Comment by etienne-lms** [comment]
Closing. The feature is being addressed through the LKML: see latest post https://lore.kernel.org/lkml/20231030084812.905549-1-etienne.carriere@foss.st.com/.

**Pull request closed without merging**